DESIGN.md

```diff
@@ -0,0 +1,756 @@
+runtime.
+
+## Submit-based operations
+
+Operations on io-uring backed resources return futures, representing the
+operation completion. The caller awaits the future to get the operation result.
+
+```rust
+let socket = my_listener.accept().await?;
+```
+
+The runtime communicates with the kernel using two single-producer,
+single-consumer queues. It submits operation requests, such as accepting a TCP
+socket, to the kernel using the submission queue. The kernel then performs the
+operation. On completion, the kernel returns the operation results via the
+completion queue and notifies the process. The `io_uring_enter` syscall flushes
+the submission queue and acquires any pending completion events. Upon request,
+this syscall may block the thread waiting for a minimum number of completion
+events. Both queues use memory shared between the process and the kernel and
+synchronize with atomic operations.
+
+Operation futures provide an asynchronous cancel function, enabling the caller
+to await on a clean cancellation.
+
+```rust
+let accept = tcp_listener.accept();
+
+tokio::select! {
+    (tcp_stream, addr) = &mut accept => { ... }
+    _ = tokio::time::sleep(Duration::from_millis(100)) => {
+        // Accept timed out, cancel the in-flight accept.
+        match accept.cancel().await {
+            Ok(_) => { ... } // operation canceled gracefully
+            Err(Completed(tcp_stream, addr)) => {
+                // The operation completed between the timeout
+                // and cancellation request.
+            }
+            Err(_) => { ... }
+        }
+    }
+}
+```
+
+In practice, operation timeouts will deserve a dedicated API to handle the
+boilerplate.
+
+```rust
+let (tcp_stream, addr) = tcp_listener
+    .accept()
+    .timeout(Duration::from_millis(100))
+    .await?
+```
+
+The cancel and timeout function are inherent methods on operation future types.
+
+If the operation's future drops before the operation completes, the runtime will
+submit a cancellation request for the operation. However, cancellation is
+asynchronous and best-effort, the operation may still complete. In that case,
+the runtime discards the operation result.
+
+The queue's single-producer characteristic optimizes for a single thread to own
+a given submission queue. Supporting a multi-threaded runtime requires either
+synchronizing pushing onto the submission queue or creating a queue pair per
+thread. The tokio-uring runtime uses an isolated thread-per-core model,
+differing from Tokio's other runtimes. Unlike Tokio's primary multi-threaded
+runtime, there is no work-stealing. Tasks remain on the thread that spawned them
+for the duration of their lifecycle. Each runtime thread will own a dedicated
+submission and completion queue pair, and operations are submitted using the
+submission queue associated with the current thread. Operation completion
+futures will not implement `Send`, guaranteeing that they remain on the thread
```

Comment:
This would have quite a large impact on the ecosystem. Is there a possibility to make the futures be `Send`, but only incur a significant performance penalty if polled on another thread?

For example, the source of the `!Send` bound appears to be the necessity to access the driver state when polling:
https://github.com/tokio-rs/tokio-uring/blob/master/src/driver/op.rs#L15

I could imagine a solution where the `Op` stores an `Arc<MagicCell<driver::Inner>>`. The idea behind `MagicCell` is that it would perform a runtime `Send` check (ie. check that the accessing thread is the same as the thread it was created on) before allowing access to the interior.

If this check fails, the `Op` would do something like this pseudocode:
```
let (tx, rx) = Proxy::new();
let original_op = mem::replace(self, rx);

original_thread.send(original_op.and_then(tx));
```

As long as tasks are fairly sticky, then the overhead would only be incurred for in-progress ops when a task is actually moved. Of course, there's the question of whether the overhead of the `Arc` and `MagicCell` would outweigh the benefits.

Reply:
It is *possible* but, part of the goal for tokio-uring, is to provide a synchronization-free option.

Related to the "ecosystem", there already are `!Send` options out there (e.g. glommio), so it already is an issue to deal with.